pages/api/signup.ts

```diff
@@ -2,51 +2,54 @@ import { NextApiRequest, NextApiResponse } from "next";
 
 export async function handler(req: NextApiRequest, res: NextApiResponse) {
     try {
-        let firstName: string = req.body.firstName();
-        let lastName: string = req.body.lastName();
-        let email: string = req.body.email();
-        let laptop: boolean = req.body.laptop();
-        let experience: string = req.body.experience();
-        let grade: number = req.body.grade();
-        let shirt: string = req.body.shirt();;
-        let diet: string = req.body.diet();;
-        let other: string = req.body.other();;
+        let firstName: string = req.body.firstName;
+        let lastName: string = req.body.lastName;
+        let email: string = req.body.email;
+        let laptop: boolean = req.body.laptop;
+        let experience: string = req.body.experience;
+        let grade: number = req.body.grade;
+        let shirt: string = req.body.shirt;
+        let diet: string = req.body.diet;
+        let other: string = req.body.other;
+        let errors = []; //grouping: names, email, long paragraphs, button choices
         const names: string[] = [firstName, lastName];
         const explains: string[] = [experience, diet, other];
-        const emailPattern = /^[a-zA-Z0-9._-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2-4}$/;
-        const namePattern = /^[a-zA-Z]{2-25}$/;
+        const choices: any[] = [laptop, grade, shirt];
+        const emailPattern = /^[a-zA-Z0-9._-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,4}$/;
+        const namePattern = /^[a-zA-Z]{2,25}$/;
         const explainPattern = /^[a-zA-Z0-9\(\)._ -]+$/;
 
-        if (grade == null || laptop == null || shirt == null) {
-            res.status(400).json({ message: "Bad Request" })
-        } else {
-            res.status(200).json({ message: "OK" })
+        for (var s of names) {
+            if (namePattern.test(s)) {
+                errors.push(res.status(200).json({ message: "OK" }));
```

Comment:
You don't need to append a success message to the errors. What you can do is say "if (!namePattern.test(s)) {...}"

The ! is to basically say the opposite. 

This would shorten the code a lot; it's also good practice to check for errors before validation.